XBee node discovery

timvandermeij opened the pull request:
This patch implements a devices view in the control panel that allows us to track the status of all XBee devices in the network. We use XBee's built-in node discovery functionality to let the ground station discover other devices in the network. When we found a device in the network, the `ND` packet contains its node identifier and address, which we store in the devices view.

We show the devices view immediately after the loading view because one needs to be sure that the XBees have joined the network before sending any waypoints to them.

@lhelwerd I already did extensive tests with this, but I do not merge this PR immediately in case you would like to look at the code too. Let me know if you have any comments or if you're fine with it, because then I'll update the PR and merge it. It's probably best to look at the overall diff and it should not be too hard to follow, but I did add some more comments in the commit messages.

timvandermeij committed:
Implement a static devices view

This view will allow us to obtain the status for each device in the
network. The device's ID, type, address and join status will be visible.

timvandermeij committed:
Implement device data storage and dynamic tree widget filling

timvandermeij committed:
Use better defaults for the tree view items

timvandermeij committed:
Implement `get_identity()` for XBee sensors

This allows us to easily obtain the identity (current status) of the
connected XBee sensor.

timvandermeij committed:
Make the ground station data dynamic

Additionally, reorder the imports in the XBee sensor physical code in
alphabetical order for readability.

timvandermeij committed:
Improve address printing and add more unit tests

timvandermeij committed:
Implement node discovery for vehicle XBees

This completes the devices view. When we first open the view, the ground
station device data is shown. The refresh button yields the information
for the other vehicle XBees using a node discovery process.

I have tested this by starting the control panel with only the ground
station inserted. After that, I inserted the two other XBee devices,
started a serial connection to them in a separate terminal (which is
required to activate them and let them join the network), and pressed
the refresh button. Both devices are found pretty quickly. It is also
nice to see that sometimes one device joins the network faster, in which
case we see the information for that device first and need another
refresh for the second device.

timvandermeij commented:
Since it's becoming increasingly hard to test this without the XBees, below is a screenshot of the view for you to see how it looks.

![device view](https://cloud.githubusercontent.com/assets/1993262/13907023/efad23e8-eee4-11e5-8cb8-9e1d5be9f652.png)

lhelwerd commented on zigbee/XBee_Sensor_Simulator.py:
```diff
@@ -126,6 +138,16 @@ def enqueue(self, packet, to=None):
+        to refresh the status of the other XBee devices.
+        """
+
+        pass
```
I guess I cannot really test it right now with the XBee simulator ;) I'm not sure how difficult it is to implement the discovery for the simulators. Could you just send a special packet/command to all the socket ports (given we know the maximum number of sensors beforehand) and respond to that with another special command? Then it would be quite similar to the physical version.

lhelwerd commented:
The code looks fine by me. I suppose you need to refresh at least once to see the sensors? I think that should work fine, although automatic detection would of course be best :wink:

The GUI takes up :100:% CPU of one core for me, but this also happens on the master branch. Perhaps we should file a separate issue for that or figure out why it happens.

lhelwerd commented:
Apparently the 100% CPU usage is specific to the `--controller-xbee-simulation` flag, since it does not seem to happen in the loading view for physical XBees at least. Perhaps you can confirm. The profiler does not give conclusive results, perhaps because it does not profile all threads. It's not a big issue since it's simulator-only, but I could delve into it more.

timvandermeij committed:
XBee sensor simulator: implement `discover` method for simulation

Previously we did nothing, but that prevents the control panel from
showing any information when the refresh button is clicked. This patch
implements a way to simulate what is being done in the physical XBee
code.

timvandermeij committed:
XBee sensor simulator: do not use 100% CPU when no data is available

When there is no data in the pipeline, we used to just continue the
loop, but by doing so we skipped the necessary loop delay a few lines
below it. This patch adds the loop delay in that path and thereby
resolves the 100% CPU usage. Note that 100% CPU was only achieved when
using the simulator component, not when using the physical XBee
component.

We did not notice this defect earlier because the runner would always
provide the simulated sensors with data, so we would never reach this
code path. For the control panel use case, however, they receive data
only sometimes, which made us notice this issue.

lhelwerd commented on zigbee/XBee_Sensor_Simulator.py:
```diff
@@ -87,6 +87,7 @@ def _loop(self):
                 try:
                     data = self._socket.recv(self.settings.get("buffer_size"))
                 except socket.error:
+                    time.sleep(self._loop_delay)
```
Nice find!

lhelwerd commented on zigbee/XBee_Sensor_Simulator.py:
```diff
@@ -146,7 +146,15 @@ def discover(self, callback):
+        # The simulator does not use XBee device discovery because it does not
+        # use the actual XBee library that provides this functionality. We
+        # simulate the process by calling the callback with the packet manually.
+        for vehicle in [1, 2]:
```
Perhaps you can make this more dynamic by using `for vehicle in range(1, self.scheduler.number_of_sensors+1)` instead. We can assume for now in the control panel view that we have 2 vehicle sensors, but we should probably solve this in a similar way later on as well.

timvandermeij replied:
Good point. I have fixed that in the new commit.

timvandermeij commented:
I have addressed your comment about the `discover` method in the simulator. I have also fixed the issue with 100% CPU usage. The cause of that issue was kind of orthogonal to this patch, but it's so small we can fix it here anyway. I could not reproduce the issue at first, but then thought it might be something with the simulator. Only shortly after that I noticed your comment, so you found it at the exact same moment as I did :-)

Let's merge this as everything is done now. Indeed we have to press the refresh button at least once, but most of the time it's only once, so that's a tiny effort ;-) Making that automatic is not trivial as we might need some kind of timer mechanism, so let's not do that for now.

timvandermeij committed:
XBee sensor simulator: make the number of vehicles dynamic

timvandermeij commented:
Let's merge this as it's done and the additional concerns have been addressed. Thanks for the review!

lhelwerd commented:
:+1:
